File: src/dev/hwlocks.cc
/** @file
 * Isa Fake Device implementation
 */

#include "dev/hwlocks.hh"

#include "base/trace.hh"
// #include "debug/IsaFake.hh"
#include "mem/packet.hh"
#include "mem/packet_access.hh"
#include "sim/system.hh"

namespace gem5
{

HWLocks::HWLocks(const Params &p)
    : BasicPioDevice(p, p.ret_bad_addr ? 0 : p.pio_size)
{
    retData8 = p.ret_data8;
    retData16 = p.ret_data16;
    retData32 = p.ret_data32;
    retData64 = p.ret_data64;
}

Tick
HWLocks::read(PacketPtr pkt)
{
    pkt->makeAtomicResponse();
    panic("***Device being accessed by cache?\n");

    // if (params().warn_access != "")
    //      warn("Device %s accessed by read to address %#x size=%d\n",
    //              name(), pkt->getAddr(), pkt->getSize());
    // if (params().ret_bad_addr) {
    //     // DPRINTF(HWLocks, "read to bad address va=%#x size=%d\n",
    //     //        pkt->getAddr(), pkt->getSize());
    //     pkt->setBadAddress();
    // } else {
        assert(pkt->getAddr() >= pioAddr && pkt->getAddr() < pioAddr + pioSize);
        // DPRINTF(HWLocks, "read  va=%#x size=%d\n",
        //        pkt->getAddr(), pkt->getSize());
        switch (pkt->getSize()) {
          case sizeof(uint64_t):
             pkt->setLE(retData64);
             break;
          case sizeof(uint32_t):
             pkt->setLE(retData32);
             break;
          case sizeof(uint16_t):
             pkt->setLE(retData16);
             break;
          case sizeof(uint8_t):
             pkt->setLE(retData8);
             break;
          default:
             if (params().fake_mem)
                 std::memset(pkt->getPtr<uint8_t>(), 0, pkt->getSize());
             else
                 panic("invalid access size! Device being accessed by cache?\n");
        }
    // }
    return pioDelay;
}

Tick
HWLocks::write(PacketPtr pkt)
{
    pkt->makeAtomicResponse();
    if (params().warn_access != "") {
        uint64_t data;
        switch (pkt->getSize()) {
          case sizeof(uint64_t):
            data = pkt->getLE<uint64_t>();
            break;
          case sizeof(uint32_t):
            data = pkt->getLE<uint32_t>();
            break;
          case sizeof(uint16_t):
            data = pkt->getLE<uint16_t>();
            break;
          case sizeof(uint8_t):
            data = pkt->getLE<uint8_t>();
            break;
          default:
            panic("invalid access size: %u\n", pkt->getSize());
        }
        warn("Device %s accessed by write to address %#x size=%d data=%#x\n",
                name(), pkt->getAddr(), pkt->getSize(), data);
    }
    if (params().ret_bad_addr) {
        // DPRINTF(HWLocks, "write to bad address va=%#x size=%d \n",
        //        pkt->getAddr(), pkt->getSize());
        pkt->setBadAddress();
    } else {
        // DPRINTF(HWLocks, "write - va=%#x size=%d \n",
        //        pkt->getAddr(), pkt->getSize());

        if (params().update_data) {
            switch (pkt->getSize()) {
              case sizeof(uint64_t):
                retData64 = pkt->getLE<uint64_t>();
                break;
              case sizeof(uint32_t):
                retData32 = pkt->getLE<uint32_t>();
                break;
              case sizeof(uint16_t):
                retData16 = pkt->getLE<uint16_t>();
                break;
              case sizeof(uint8_t):
                retData8 = pkt->getLE<uint8_t>();
                break;
              default:
                panic("invalid access size!\n");
            }
        }
    }
    return pioDelay;
}

// HWLocksDriver*
// HWLocksDriverParams::create()
// {
//     return new HWLocksDriver(this);
// }

} // namespace gem5

File: src/dev/HWLocks.py
from m5.params import *
from m5.objects.Device import BasicPioDevice
from m5.objects.Process import EmulatedDriver

class HWLocks(BasicPioDevice):
    type = 'HWLocks'
    cxx_header = "dev/hwlocks.hh"
    cxx_class = 'gem5::HWLocks'

    devicename = Param.String("hwlocks", "Name of device to error on")

    # parameters copied from Device.py for IsaFake
    pio_size = Param.Addr(0x8, "Size of address range")
    ret_data8 = Param.UInt8(0xFF, "Default data to return")
    ret_data16 = Param.UInt16(0xFFFF, "Default data to return")
    ret_data32 = Param.UInt32(0xFFFFFFFF, "Default data to return")
    ret_data64 = Param.UInt64(0xFFFFFFFFFFFFFFFF, "Default data to return")
    ret_bad_addr = Param.Bool(False, "Return pkt status bad address on access")
    update_data = Param.Bool(False, "Update the data that is returned on writes")
    warn_access = Param.String("", "String to print when device is accessed")
    fake_mem = Param.Bool(False,
      "Is this device acting like a memory and thus may get a cache line sized req")

class HWLocksDriver(EmulatedDriver):
    type = 'HWLocksDriver'
    cxx_header = "dev/hwlocks.hh"
    filename = "hwlocks"

    hardware = Param.HWLocks("The hwlocks hardware")